ath79: add support for TP-Link EAP245-v3

svanheule opened the pull request:
This pull request provides the necessary pieces to support the TP-Link EAP245v3, a 3x3 802.11ac wave-2 ceiling mount access point with PoE and ethernet pass-through.

The following patches are provided:
* New MTD splitter for firmware partitions that start with an elf loader, followed by the rootfs,
* Small patch that enables the new splitter for the ath79 target,
* Small patch for the safeloader image builder that clarifies the data length provided in the soft-version partition,
* The "additional hardware version" TP-Link now stores in the soft-version partition, implemented as a compatibility level,
* And finally the patch that provides support for the EAP245v3.

The TP-Link bootloader will only boot an ELF loader without manual intervention, so to provide a single (dynamic) firmware partition, the new ELF splitter was implemented.

svanheule committed:
ath79: add support for TP-Link EAP245-v3

TP-Link EAP245 v3 is an AC1750 (802.11ac Wave-2) ceiling mount access
point. UART access (for debricking) requires non-trivial soldering.

Specifications:
* SoC: QCA9563 (CPU/DDR/AHB @ 775/650/258 MHz)
* RAM: 128MiB
* Flash: 16MiB SPI-NOR
* Wireless 2.4GHz (SoC): b/g/n 3x3
* Wireless 5GHz (QCA9982): a/n/ac 3x3 with MU-MIMO
* Ethernet (QCA8337N switch): 2× 1GbE, ETH1 (802.3at PoE) and ETH2
* Green and amber status LEDs
* Reset switch (GPIO, available for failsafe)

Flashing instructions:
All recent firmware versions (latest is 2.20.0), can disable firmware
signature verification and use a padded firmware file to flash OpenWrt:
* ssh into target device and run `cliclientd stopcs`
* upload factory image via web interface

The stopcs-method is supported from firmware version 2.3.0. Earlier
versions need to be upgraded to a newer stock version before flashing
OpenWrt.

Factory images for these devices are RSA signed by TP-Link. While the
signature verification can be disabled, the factory image still needs to
have a (fake) 1024 bit signature added to pass file checks.

Debricking instructions:
You can recover using u-boot via the serial port:
* Serial port is available from J3 (1:TX, 2:RX, 3:GND, 4:3.3V)
* Bridge R237 to connect RX, located next to J3
* Bridge R225 to connect TX, located inside can on back-side of board
* Serial port is 115200 baud, 8n1, interrupt u-boot by holding ctrl+B
* Upload initramfs with tftp and upgrade via OpenWrt

Device mac addresses:
Stock firmware has the same mac address for 2.4GHz wireless and
ethernet, 5GHz is incremented by one. The base mac address is stored in
the 'default-mac' partition (offset 0x90000) at an offset of 8 bytes.
ART blobs contain no mac addresses.
From OEM ifconfig:
    ath0      Link encap:Ethernet  HWaddr 74:..:E2
    ath10     Link encap:Ethernet  HWaddr 74:..:E3
    br0       Link encap:Ethernet  HWaddr 74:..:E2
    eth0      Link encap:Ethernet  HWaddr 74:..:E2

Signed-off-by: Sander Vanheule <sander@svanheule.net>

adrianschmutzler commented on target/linux/ath79/dts/qca9563_tplink_eap245-v3.dts:
```diff
@@ -0,0 +1,172 @@
+		led_orange: led_orange {
+			label = "tp-link:orange:failsafe";
+			gpios = <&gpio 9 GPIO_ACTIVE_HIGH>;
+		};
```
I'd refer to these as status LEDs, leading to the following:
```
		led_status_green: status_green {
			label = "tp-link:green:status";
			gpios = <&gpio 7 GPIO_ACTIVE_HIGH>;
			default-state = "on";
		};

		led_status_orange: status_amber {
			label = "tp-link:amber:status";
			gpios = <&gpio 9 GPIO_ACTIVE_HIGH>;
		};
```

The default-state will have the LED on earlier in the boot process, i.e. before scripts like the one evaluating led-aliases will work.

svanheule replied:
I didn't really like the 'failsafe' name for the amber LED either, but couldn't think of anything more appropriate. Thanks for the suggestion.

adrianschmutzler commented on target/linux/ath79/dts/qca9563_tplink_eap245-v3.dts:
```diff
@@ -0,0 +1,172 @@
+
+&gpio {
+	status = "okay";
+};
```
gpio is not disabled, so this can be dropped.

adrianschmutzler commented on target/linux/ath79/dts/qca9563_tplink_eap245-v3.dts:
```diff
@@ -0,0 +1,172 @@
+			};
+
+			partition@f30000 {
+				label = "mutil-log";
```
really "mutil"?

svanheule replied:
Yes, "mutil". I don't know if this refers to anything in particular, but from TP-Link's partition names I can only assume little energy was spent on ensuring their correctness.

From the bootlog:
```
[    0.460000] 11 cmdlinepart partitions found on MTD device spi0.0
[    0.470000] Creating 11 MTD partitions on "spi0.0":
[    0.480000] 0x000000000000-0x000000040000 : "factory-boot"
[    0.480000] 0x000000040000-0x000000080000 : "u-boot"
[    0.490000] 0x000000080000-0x000000090000 : "pation-table"
...
[    0.540000] 0x000000f30000-0x000000fb0000 : "mutil-log"
[    0.550000] 0x000000fb0000-0x000000ff0000 : "oops"
```

From the partition-table/pation-table partition:
```
partition factroy-boot base 0x00000 size 0x40000
partition fs-uboot base 0x40000 size 0x40000
partition partition-table base 0x80000 size 0x02000
...
partition mutil-log base 0xf30000 size 0x80000
partition oops base 0xfb0000 size 0x40000
```

adrianschmutzler commented on target/linux/ath79/dts/qca9563_tplink_eap245-v3.dts:
```diff
@@ -0,0 +1,172 @@
+				reg = <0xfb0000 0x040000>;
+				read-only;
+			};
+
```
drop empty line here.

svanheule replied:
Also dropped another empty line after the led-block (IIRC)

adrianschmutzler commented on target/linux/ath79/dts/qca9563_tplink_eap245-v3.dts:
```diff
@@ -0,0 +1,172 @@
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&info 0x8>;
+};
+
```
drop empty line at EOF.

adrianschmutzler commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -1291,6 +1291,37 @@ static struct device_info boards[] = {
 
+	/** Firmware layout for the EAP245 v3 */
+	{
+		.id     = "EAP245-V3",
```
I've never seen this without .vendor defined, though empty. Does it still work and can all the empty vendor statements by dropped?

svanheule replied:
TP-Link's firmware files for this device just contain the file length (minus the signature) and md5 checksum at the start of the file. Then it's FF until the partition list at 0x1014. This is true for firmware files for:
* EAP245 v1 and v3
* EAP225 v1, v2 and v3
If .vendor is not defined, then the produced factory image also contains FF after the md5 sum, replicating the TP-Link format.

As for empty-string .vendor definitions, I can't really tell if they would work. The Archer C60-v3's .vendor is emtpy, so the factory images have {00 00 00 00 00] after the checksum. TP-Link's latest image for this device however, contains a header of 0x28c bytes ("fw-type:Cloud\n[binary data]"). If these factory images currently work, then I would assume removing the empty .vendor strings wouldn't hurt, but I cannot test this as I don't have such a device.

adrianschmutzler commented:
You might increase the chance of somebody looking at your mtdsplit patch by sending it separately to the openwrt-devel list.

svanheule commented on target/linux/ath79/dts/qca9563_tplink_eap245-v3.dts:
```diff
@@ -19,6 +19,7 @@
 		led-failsafe = &led_status_amber;
 		led-running = &led_status_green;
 		led-upgrade = &led_status_amber;
+		label-mac-device = &eth0;
```
I noticed other devices also include this alias, so I decided to add it here too. If just adding it here has no real use, feel free to drop this additional patch.

adrianschmutzler replied:
This alias has the use that somebody can easily retrieve the MAC address printed on the label via the function get_mac_label:
https://openwrt.org/docs/guide-developer/mac.address#using_the_label_mac_address

This might be helpful e.g. for downstream networks (Freifunk) that want to identify devices in large networks by that.

svanheule commented:
> You might increase the chance of somebody looking at your mtdsplit patch by sending it separately to the openwrt-devel list.

Just sent out the emails. I think it can be useful for other devices too.

adrianschmutzler commented on target/linux/ath79/dts/qca9563_tplink_eap245-v3.dts:
```diff
@@ -0,0 +1,166 @@
+
+	chosen {
+		bootargs = "console=ttyS0,115200n8";
+	};
```
Just pushed a patch to master. chosen node can be dropped here.

svanheule commented:
I've squashed my extra patches and rebased to the current master, to ensure the merged chosen/bootargs are used.

With the issues on the openwrt-devel list, I think my mtdsplit patch may not have gotten too many views. It showed up wrapped and with "no subject", and doesn't seem to have made it into to patchworks. Should I resend my patch?

adrianschmutzler commented:
> Should I resend my patch?

Yes, if you don't see it in patchwork: https://patchwork.ozlabs.org/project/openwrt/list/

svanheule commented:
I found a way to upload unsigned images to this device with a different image format: https://github.com/svanheule/tplink-fw-3rd

The factory image script currently just reworks a sysupgrade file, but should be converted into a proper image builder. I'll try to send to the mailing list to discuss the best way forward.

PolynomialDivision commented:
Could this also work with an TP-Link EAP 225 v3?

svanheule commented:
> Could this also work with an TP-Link EAP 225 v3?

If you are referring to the image format, I'm afraid it won't. The EAP 245 v3 seems to use a different firmware version than the EAP 225 v3. More specifically, the libnvrammanager.so file that contains the support for this image type, isn't present in the EAP 225 v3's firmware. See also the [EAP245 thread](https://forum.openwrt.org/t/adding-openwrt-support-for-tp-link-eap245/57583) for details and updates.

svanheule committed:
kernel: mtdsplit: support ELF loader splitting

To parse the ELF kernel loader, a small ELF parser is used that can
handle both ELF32 or ELF64 class loaders. The splitter assumes that the
kernel is always located before the rootfs, whether it is embedded in
the loader or not. If the kernel is located after the rootfs on the
firmware partition, then the rootfs splitter will include it in the
dynamically created rootfs_data partition and the kernel will be
corrupted.

The kernel image is preferably embedded inside the ELF loader, so the
end of the loader equals the end of the kernel partition. This is due to
the way mtd_find_rootfs_from searches for the the rootfs:
- if the kernel image is embedded in the loader, the appended rootfs may
  follow the loader immediately, within the same erase block.
- if the kernel image is not embedded in the loader, but placed at some
  offset behind the loader (OKLI-style loader), the rootfs must be
  aligned to an erase-block after the loader and kernel image.

In case section header table is empty, determine the elf loader size by
finding the end of the last segment, as defined by the program header
table.

Signed-off-by: Sander Vanheule <sander@svanheule.net>

svanheule commented:
I've made some big changes:
* Implemented more generic ELF splitter instead of OKLI splitter.
 This introduces a new mtdsplit source file and some options that need to be switched on for the ath79 kernel. It also simplifies the device directives. The OKLI scanner used the default search step size, so that might have caused restrictions in the future.
* Found a better way to flash the factory image. Soldering is no longer required unless you need to de-brick.
 I've tested this with an EAP225v3 firmware (running on EAP245v1 hardware). I found this method in the EAP245v3 software, but could only use the other device to test the flashing.

Would you like me to send the elf splitter to the mailing list? I've tried relaxing my DMARC configuration, but haven't gotten any patch into patchwork yet. My last attempt for the OKLI splitter patch could have been caused by DNS cacheing though.

ynezz commented:
@rmilecki another mtdsplitter change, also sent for review to the list https://patchwork.ozlabs.org/project/openwrt/patch/20200708134343.79976-2-sander@svanheule.net/

svanheule committed:
ath79: enable elf mtd splitter

Enabled the ELF firmware partition splitter 4.19 and 5.4 in preparation
for the TP-Link EAP245v3 device support.

Signed-off-by: Sander Vanheule <sander@svanheule.net>

svanheule commented:
I've sent an updated version of the mtd splitter patch to the mailing list and rebased on top of that.

svanheule commented:
My apologies for the noise today with another push.
Changes:
* Dropped `Device/tplink-safeloader-elf` from the makefiles, putting everything into the device directives. Since this is currently the only device with this kind of image, this seemed better to me for now.
* Updated commit message:
  * Simplified flashing instructions
  * Added mac address info

svanheule committed:
firmware-utils/tplink-safeloader: soft-version magic is data length

The soft-version partition actually contains a header and trailing data:
* header: {data length, [zero]}
* data: {version, bcd encoded date, revision}

The data length is currently treated as a magic number, but should
contain the length of the partition data.

This header is also present the following partitions (non-exhaustive):
* string-based soft-version
* support-list

Signed-off-by: Sander Vanheule <sander@svanheule.net>

svanheule committed:
firmware-utils/tplink-safeloader: add compat level

TP-Link has introduced a compatibility level to prevent certain
downgrades. This information is stored in the soft-version partition,
changing the data length from 0xc to 0x10.

The compatibility level doesn't change frequently. For example, it has
the following values for the EAP245v3 (released 2018-Q4):
* FW v2.2.0  (2019-05-30): compat_level=0
* FW v2.3.0  (2019-07-31): compat_level=0
* FW v2.3.1  (2019-10-29): compat_level=1
* FW v2.20.0 (2020-04-23): compat_level=1

Empty flash values (0xffffffff) are interpreted as compat_level=0.
If a firmware upgrade file has a soft-version block without
compatibility level (data length < 0x10), this is also interpreted as
compat_level=0.

By including a high enough compatibility level in factory images, stock
firmware can be convinced to accept the image. A compatibility level
aware firmware will keep the original value.

Example upgrade log of TP-Link EAP245v3 FWv2.3.0 to FWv2.20.0:
    [NM_Debug](nm_fwup_verifyFwupFile) 02073: curSoftVer:2.3.0 Build
        20190731 Rel. 51932,newSoftVer:2.20.0 Build 20200423 Rel. 36779
    ...
    AddiHardwareVer check: NEW(0x1) >= CUR(0x0), Success.
    ...
    [NM_NOTICE](updateDataToNvram) 00575: Restore old additionalHardVer:
    0x0.(new 0x1)
    [NM_NOTICE](updateDataToNvram) 00607: PTN 07: name = soft-version,
        base = 0x00092000, size = 0x00000100 Bytes, upDataType = 1,
        upDataStart = 7690604b, upDataLen = 00000018
    [NM_Debug](updateDataToNvram) 00738: PTN 07: write bytes = 000002eb

Other firmware upgrades have been observed to modify the compabitility
stored level (e.g. TP-Link EAP225-Outdoor FWv1.4.1 to FWv1.7.0).
Therefore, it seems to be the safest option to set the OpenWrt
compatibility level to the highest known value instead of the highest
possible value (0xfffffffe), to ensure users do not get unexpectedly
refused firmware upgrades when using a device reverted back to stock.

To remain compatible with existing devices and not produce different
images, the image builder doesn't store a compatibility level if it is
zero.

Signed-off-by: Sander Vanheule <sander@svanheule.net>

svanheule commented:
Since I was testing firmware upgrades on different devices, I missed a few errors still:
* Safeloader support-list is mixed case, not all upper case
* Partition names need to match between factory image and names stored in flash, even if the partition isn't upgraded (i.e. 'fs-uboot' cannot be named 'u-boot' here)

Because stock FW versions >2.3.0 have a compatibilitiy level ("additionalHardwareVersion") of 1, OpenWrt factory images now also require to be at least this value. This is for the cases where a user purchases a new device with FW version 2.3.1 or higher. The two extra tplink-safeloader commits implement the necessary code to produce compatible images.

svanheule commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -77,6 +77,7 @@ struct device_info {
 	const char *support_list;
 	char support_trail;
 	const char *soft_ver;
+	uint32_t soft_ver_compat_level;
```
The compat_level-changing upgrade described in the commit message (EAP225-Outdoor FWv1.4.1 to FWv1.7.0) was necessary to enable flashing of OpenWrt with checksum disabling (v1.4.1 didn't have the required `cliclientd` utility). 

Since all firmwares I know that have the `cliclientd` utility also appear to preserve the original compatibility level, it *might* be possible to just use 0xfffffffe for OpenWrt and call it a day. However, I don't feel very confident about this and would rather stay close to TP-Link stock images with a low, but device-specific compat_level.

svanheule commented:
I had a fixup for the first tplink-safeloader patch in the second patch, hence another push. No code changes however, and sorry for the extra noise!

adrianschmutzler commented:
Just had checkpatch.pl running on these patches, this is the output FYI:

```
WARNING: line over 80 characters
#191: FILE: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_elf.c:130:
+               pr_debug("invalid ELF magic %08x\n", be32_to_cpu(hdr.ident.magic));

WARNING: line over 80 characters
#197: FILE: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_elf.c:136:
+               loader_size = hdr.elf32.shoff + hdr.elf32.shnum * hdr.elf32.shentsize;

WARNING: line over 80 characters
#200: FILE: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_elf.c:139:
+               loader_size = hdr.elf64.shoff + hdr.elf64.shnum * hdr.elf64.shentsize;

total: 0 errors, 3 warnings, 208 lines checked
```
for Applying: kernel: mtdsplit: support ELF loader splitting

```
WARNING: line over 80 characters
#38: FILE: tools/firmware-utils/src/tplink-safeloader.c:2033:
+       /* Size of the partition contents, without 8 byte header and padding byte */

total: 0 errors, 1 warnings, 17 lines checked
```
for Applying: firmware-utils/tplink-safeloader: soft-version magic is data length

```
ERROR: open brace '{' following function declarations go on the next line
#80: FILE: tools/firmware-utils/src/tplink-safeloader.c:2020:
+static struct image_partition_entry make_soft_version(struct device_info *info, uint32_t rev) {

WARNING: line over 80 characters
#85: FILE: tools/firmware-utils/src/tplink-safeloader.c:2025:
+       struct image_partition_entry entry = alloc_image_partition("soft-version", part_len+1);

WARNING: suspect code indent for conditional statements (8, 12)
#94: FILE: tools/firmware-utils/src/tplink-safeloader.c:2052:
+       if (info->soft_ver_compat_level > 0)
+           *(uint32_t *)(entry.data + sizeof(struct soft_version)) = htonl(info->soft_ver_compat_level);

WARNING: line over 80 characters
#95: FILE: tools/firmware-utils/src/tplink-safeloader.c:2053:
+           *(uint32_t *)(entry.data + sizeof(struct soft_version)) = htonl(info->soft_ver_compat_level);

total: 1 errors, 3 warnings, 47 lines checked
```
for Applying: firmware-utils/tplink-safeloader: add compat level

```
WARNING: line over 80 characters
#307: FILE: tools/firmware-utils/src/tplink-safeloader.c:1304:
+               /** Merged os-image/file-system into dynamic firmware partition */

ERROR: space required before the open parenthesis '('
#336: FILE: tools/firmware-utils/src/tplink-safeloader.c:2423:
+       } else if(strcasecmp(info->id, "EAP245-V3") == 0) {

WARNING: line over 80 characters
#337: FILE: tools/firmware-utils/src/tplink-safeloader.c:2424:
+               const char mdat[10] = {0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x01, 0x01};

total: 1 errors, 2 warnings, 251 lines checked
```
for Applying: ath79: add support for TP-Link EAP245-v3
(for the last one I'd not be too strict, as nobody cares anyway as it seems)

svanheule commented:
> for Applying: kernel: mtdsplit: support ELF loader splitting

Fixed.

> for Applying: firmware-utils/tplink-safeloader: soft-version magic is data length
> for Applying: firmware-utils/tplink-safeloader: add compat level
> for Applying: ath79: add support for TP-Link EAP245-v3
> (for this one I'd not be too strict, as nobody cares anyway as it seems)

I've fixed most of these lines. Though as you noticed, tplink-safeloader.c does not adhere to the general coding guidelines. Open braces '{' on the same line as the function declaration is just the style that is used throughout the file, and long lines (>80 chararacters) are also everywhere.

This leaves the following:
* for "firmware-utils/tplink-safeloader: add compat level"
```
ERROR: open brace '{' following function declarations go on the next line
#80: FILE: tools/firmware-utils/src/tplink-safeloader.c:2027:
+static struct image_partition_entry make_soft_version(struct device_info *info, uint32_t rev) {

total: 1 errors, 0 warnings, 49 lines checked
```

* for "ath79: add support for TP-Link EAP245 v3"
```
WARNING: line over 80 characters
#342: FILE: tools/firmware-utils/src/tplink-safeloader.c:2433:
+		const char mdat[10] = {0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x01, 0x01};

total: 0 errors, 1 warnings, 256 lines checked
```


If you would like tplink-safeloader.c to be cleaned up, I can give that a try after this PR.
```
$ scripts/checkpatch.pl <(diff -u /dev/null tools/firmware-utils/src/tplink-safeloader.c)
...
total: 27 errors, 112 warnings, 2903 lines checked
```

Another change in the last push, is that I've split up the Device-entry in the makefile into two entries:
* Generic Device/tplink-eap2x5 that implements the requirements for this device and the ones on the mailing list
* Specific Device/tplink_eap245-v3 that implements the details of this device

svanheule commented:
Reworded my commit messages a bit, no code changes.

svanheule commented on target/linux/ath79/dts/qca9563_tplink_eap245-v3.dts:
```diff
@@ -0,0 +1,163 @@
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+		phy-mode = "sgmii";
```
@blocktrron Reading the sources for the ar71xx ethernet drive, I suspect that specifying phy-mode in this node is not useful since it is never really used. The phy-mode at line 153 does appear to be used by the driver. Looking at other devices, I couldn't really see a specific pattern in where to use phy-mode. Would you be able to clarify?

PolynomialDivision replied:
Isn't that used in the mac layer?
There is not much documentation about it. :/
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/net/ethernet-controller.yaml#n66

svanheule replied:
I don't remember finding that document when I looked into this earlier. Thinking about it, it makes sense though. Both the mac and the phy have to use the same link type, so I has to be defined for both (if I understand this correctly).

svanheule commented:
I've rebased on top of the [mtdsplitter patch](https://patchwork.ozlabs.org/project/openwrt/patch/20200819113607.34273-1-sander@svanheule.net/) currently posted on the mailing list.

svanheule commented:
I've put the options in "ath79: enable elf mtd splitter" in alphabetical order, so if kernel config gets regenerated, the options don't change in ordering.

stintel commented:
Merged, thanks!

PolynomialDivision commented:
@svanheule thanks for your work! I eagerly awaited this moment. xD
